src/world/Action/EffectBuilder.cpp

```diff
@@ -31,62 +31,229 @@ uint64_t EffectBuilder::getResultDelayMs()
 {
-  auto result = getResult( target );
-  assert( result );
+  auto resultList = getResultList( effectTarget );
+  assert( resultList );
 
-  result->heal( amount, severity );
+  EffectResultPtr nextResult = make_EffectResult( healingTarget, getResultDelayMs() );
+  nextResult->heal( amount, severity, flag );
+  resultList->push_back( std::move( nextResult ) );
 }
 
-void EffectBuilder::damageTarget( Entity::CharaPtr& target, uint32_t amount, Common::ActionHitSeverityType severity )
+void EffectBuilder::restoreMP( Entity::CharaPtr& target, Entity::CharaPtr& restoringTarget, uint32_t amount, Common::ActionEffectResultFlag flag )
 {
-  auto result = getResult( target );
-  assert( result );
+  auto resultList = getResultList( target );
+  assert( resultList );
 
-  result->damage( amount, severity );
+  EffectResultPtr nextResult = make_EffectResult( restoringTarget, getResultDelayMs() ); // restore mp source actor
+  nextResult->restoreMP( amount, flag );
+  resultList->push_back( std::move( nextResult ) );
+}
+
+void EffectBuilder::damage( Entity::CharaPtr& effectTarget, Entity::CharaPtr& damagingTarget, uint32_t amount, Common::ActionHitSeverityType severity, Common::ActionEffectResultFlag flag )
+{
+  auto resultList = getResultList( effectTarget );
+  assert( resultList );
+
+  EffectResultPtr nextResult = make_EffectResult( damagingTarget, getResultDelayMs() );
+  nextResult->damage( amount, severity, flag );
+  resultList->push_back( std::move( nextResult ) );
+}
+
+void EffectBuilder::startCombo( Entity::CharaPtr& target, uint16_t actionId )
+{
+  auto resultList = getResultList( target );
+  assert( resultList  );
+
+  EffectResultPtr nextResult = make_EffectResult( target, 0 );
+  nextResult->startCombo( actionId );
+  resultList->push_back( std::move( nextResult ) );
+}
+
+void EffectBuilder::comboSucceed( Entity::CharaPtr& target )
+{
+  auto resultList = getResultList( target );
+  assert( resultList  );
+
+  EffectResultPtr nextResult = make_EffectResult( target, 0 );
+  nextResult->comboSucceed();
+  resultList->push_back( std::move( nextResult ) );
 }
 
 void EffectBuilder::buildAndSendPackets()
 {
+  auto targetCount = m_resolvedEffects.size();
   Logger::debug( "EffectBuilder result: " );
-  Logger::debug( "Targets afflicted: {}", m_resolvedEffects.size() );
+  Logger::debug( "Targets afflicted: {}", targetCount );
 
-  for( auto it = m_resolvedEffects.begin(); it != m_resolvedEffects.end(); )
+  auto globalSequence = m_sourceChara->getCurrentTerritory()->getNextEffectSequence();
+
+  do // we want to send at least one packet even nothing is hit so other players can see
   {
-    auto result = it->second;
-    Logger::debug( " - id: {}", result->getTarget()->getId() );
+    auto packet = buildNextEffectPacket( globalSequence );
+    m_sourceChara->sendToInRangeSet( packet, true );
```

Comment:
the only problem with this is that `sourceSequence` can only be set for the caster, so you'd need to copy the packet here and zero that field when sending it to everyone else

Reply:
that's not the case. sourceSequence is definitely counting up in other players' Effect as well. I can send a capture of me watching some random person hitting a fate monster if you want to see

Reply:
hmm, alright, seem to recall that being the case previously but that was also more than a year ago when I actively looked into that packet so it could have changed or I'm remembering that wrong too

Reply:
I did some tests there if you sent a sourceSequence = 0, or anything less than previous sent value for the same actor, the effect will not delay correctly, it will show effect immediately rather than sync with "the actor's weapon hitting the target". Maybe the client assumes it just lag and the effect is something already happened.

Reply:
yeah, the sequence is used to basically sync up the animation on the caster, for example:

1. cast action a with seq = 2
2. cast succeeds, client automagically starts anim
3. client receives effect packet with caster seq = 3
4. client invalidates and starts playing new anim for seq 3 because to the client, it's not the same cast and effect packets take precedence on what is shown and will cancel out effects/anims if it is set in an effect packet

alternatively:

1. cast action a with seq = 2
2. cast succeeds, client automagically starts anim
3. client receives effect packet with caster seq = 2
4. seq matches most recent cast seq, current anim continues to play out with existing timing as its the same cast and nothing changes other than damage dealt being set to appear